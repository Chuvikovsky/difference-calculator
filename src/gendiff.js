export default (data1, data2) => {
  const keys1 = Object.keys(data1);
  const keys2 = Object.keys(data2);
  const commonKeys = keys2.filter((k) => keys1.includes(k));
  const filteredKeys2 = keys2.flatMap((k) => (commonKeys.includes(k) ? [] : k));
  const sortedKeys = [...keys1, ...filteredKeys2].sort();
  return sortedKeys.reduce((acc, k) => {
    if (commonKeys.includes(k) && data1[k] === data2[k]) {
      return { ...acc, [`  ${k}`]: data1[k] };
    }
    if (commonKeys.includes(k) && data1[k] !== data2[k]) {
      return { ...acc, [`- ${k}`]: data1[k], [`+ ${k}`]: data2[k] };
    }
    if (keys1.includes(k)) {
      return { ...acc, [`- ${k}`]: data1[k] };
    }
    if (keys2.includes(k)) {
      return { ...acc, [`+ ${k}`]: data2[k] };
    }
    return acc;
  }, {});
};
