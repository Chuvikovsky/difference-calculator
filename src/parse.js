const getFileExt = (fileName) => fileName.split('.').at(-1).toLowerCase();

export default getFileExt;
