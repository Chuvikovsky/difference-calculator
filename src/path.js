import { cwd } from 'node:process';
import path from 'node:path';

const createAbsolutePath = (pathToFile) => path.resolve(cwd(), pathToFile);

export default createAbsolutePath;
