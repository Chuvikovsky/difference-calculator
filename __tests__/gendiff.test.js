import { fileURLToPath } from 'url';
import { dirname } from 'path';
import path from 'node:path';
import gendiff from '../src/gendiff.js';
import parse from '../src/parsers.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const PathToFix = (file) => path.join(__dirname, '..', '__fixtures__', file);

test('flat json', () => {
  const path1 = PathToFix('file1.json');
  const path2 = PathToFix('file2.json');

  const diffObj = gendiff(parse(path1), parse(path2));
  const result = Object.entries(diffObj).reduce(
    (acc, val) => [...acc, `  ${val[0]}: ${val[1]}`],
    [],
  );
  expect(['{', ...result, '}'].join('\n')).toMatchSnapshot();
});

test('flat yaml', () => {
  const path1 = PathToFix('file1.yml');
  const path2 = PathToFix('file2.yml');

  const diffObj = gendiff(parse(path1), parse(path2));
  const result = Object.entries(diffObj).reduce(
    (acc, val) => [...acc, `  ${val[0]}: ${val[1]}`],
    [],
  );
  expect(['{', ...result, '}'].join('\n')).toMatchSnapshot();
});
