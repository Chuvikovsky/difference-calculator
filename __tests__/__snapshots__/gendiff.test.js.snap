// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flat json 1`] = `
"{
  - follow: false
    host: hexlet.io
  - proxy: 123.234.53.22
  - timeout: 50
  + timeout: 20
  + verbose: true
}"
`;

exports[`flat yaml 1`] = `
"{
  - follow: false
    host: hexlet.io
  - proxy: 123.234.53.22
  - timeout: 50
  + timeout: 20
  + verbose: true
}"
`;
